### PR TITLE
fix(Core/Spell): solve issue with Life Tap All Ranks

### PR DESCRIPTION
**Pull request by Stifler82**
## Changes Proposed:
-  I fixed the fizzle, at 20% it get activates (approx) depends on the level / HP (Life Tap Rank 1)
-  I fixed the amount of life subtracted from the spell to convert to mana, now each rank of "Life Tap" subtracts the same amount of life it adds to mana previously it subtracted 34 life points, now it subtracts 27 points(as it should does with Rank 1) 
and in order also the other Ranks:
Rank 2 | 66
Rank 3 | 132
Rank 4 | 215
Rank 5 | 306
Rank 6 | 827
Rank 7 | 1124
Rank 8 | 2000

[EDIT]
- added SQL query in the "spell_dbc" table of Life Tap (All Ranks)
Fix mana returned and correction Tooltip

## Issues Addressed:
- Closes 

## Tests Performed:
-Compiled without errors
-Tested in game and working
https://youtu.be/5X3usUTZiJI **(New! Video! English Version)** **(Update 05/February/2021 )**
-Tested on Windows 10 (64)

## How to Test the Changes:

- Check the amount of life subtracted from the spell and the fizzle.
- Check the amount of mana returned
- https://youtu.be/5X3usUTZiJI My Test and Investigation about the Spell.

## Known Issues and TODO List:

- [x] To complete this fix I also had to patch the Game Client to make the fix made appear "Visually" https://github.com/azerothcore/azerothcore-wotlk/discussions/4462
This is the post with the suggestion I gave for the Game Client Fixes
- [ ] 

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by pklloveyou** [comment]
Skills are wrong for level 80 characters

**Comment by Stifler82** [comment]
> Skills are wrong for level 80 characters

Obviously, because I only fixed Rank 1.
I apologize, my fault, I hadn't specified it in the title, I did it only in the description..

File: data/sql/updates/pending_db_world/rev_1612257567543292300.sql [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
...0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '-1', '0', '0', '3', '0', '0', '1', '0', '0', '0', '0', '0', '26', '0', '0', '0', '0', '0', '1', '0', '0', '0', '0', '0', '28', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '1225', '0', '208', '0', '0', 'Life Tap', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '16712190', 'Rank 1', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '16712190', 'Converts ${$m1} health into ${$m1*$<mult>+$SPS*.5*$<mult>} mana.  Spell power increases the amount of mana returned.\r\n', 'Converts ${$m1} health into ${$m1*$<mult>+$SPS*.5*$<mult>} mana.  Spell power increases the amount of mana returned.', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '16712190', '', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '0', '16712190', '0', '133', '1500', '0', '5', '262144', '0', '0', '0', '1', '1', '0', '1065350000', '1065350000', '1065350000', '0', '0', '0', '0', '0', '0', '32', '0', '0', '0', '0', '0', '0', '175', '0');
```
Comment (FrancescoBorzi) [inline data/sql/updates/pending_db_world/rev_1612257567543292300.sql:17]: you don't need multiple DELETE/INSERT statements, you can just use one as explained in: https://www.azerothcore.org/wiki/Dealing-with-SQL-files

File: src/server/scripts/Spells/spell_warlock.cpp [hunk 2]
```diff
@@ -967,7 +967,7 @@ class spell_warl_life_tap : public SpellScriptLoader
 
         SpellCastResult CheckCast()
         {
-            if ((int32(GetCaster()->GetHealth()) > int32(GetSpellInfo()->Effects[EFFECT_0].CalcValue() + (6.3875 * GetSpellInfo()->BaseLevel) + (GetCaster()->GetStat(STAT_SPIRIT) * 1.5f) + 1.0f )))
+            if ((int32(GetCaster()->GetHealth()) > int32(GetSpellInfo()->Effects[EFFECT_0].CalcValue() + (3.1 * GetSpellInfo()->BaseLevel) + 0.9f )))
```
Comment (FrancescoBorzi) [inline src/server/scripts/Spells/spell_warlock.cpp:970]: shall we move these magic numbers inside constants? especially the `0.9f` as it's repeated

Comment (Stifler82) [reply src/server/scripts/Spells/spell_warlock.cpp:970]: `damage int32 = GetEffectValue () + 0.9f;`

**Specifies the amount of life subtracted.**

Comment (FrancescoBorzi) [reply src/server/scripts/Spells/spell_warlock.cpp:970]: I meant the `0.9f`

File: data/sql/updates/pending_db_world/rev_1612257567543292300.sql (hunk 1 above)
Comment (Kitzunu) [reply data/sql/updates/pending_db_world/rev_1612257567543292300.sql]: Also dont use ' around numbers

**Comment by Kitzunu** [comment]
Where are these values taken from?

**Comment by Stifler82** [comment]
> Where are these values taken from?
https://www.collinsdictionary.com/dictionary/english/common-sense

**Comment by Voxstrasza** [comment]
> 
> 
> > Where are these values taken from?
> > [Common Sense](https://www.collinsdictionary.com/dictionary/english/common-sense)

hmm if it's common sense then why did you open an issue suggesting we add MPQ files to change tooltips based on this PRs changes?

sounds like you pulled these values out of your arse

**Comment by Kitzunu** [comment]
Well, it has to be based on Blizzlike values, so somewhere you must have come across some comment or video that shows that these values are correct?

**Comment by Stifler82** [comment]
> Well, it has to be based on Blizzlike values, so somewhere you must have come across some comment or video that shows that these values are correct?

> > > Where are these values taken from?
> > > [Common Sense](https://www.collinsdictionary.com/dictionary/english/common-sense)
> 
> hmm if it's common sense then why did you open an issue suggesting we add MPQ files to change tooltips based on this PRs changes?
> 
> sounds like you pulled these values out of your arse

As it seems neither of you have seen my video where I show the Fix and the percentages and test it too..
So it seems obvious to me that both of you are asking for explanations. maybe, if you tried to watch my video you would realize that I am right, who knows..

**Comment by Kitzunu** [comment]
But how would I know the video would explain it? It is not stated anywhere. Anyhow, I can watch it later :)

**Comment by Voxstrasza** [comment]
I watched the video and did some research, the tooltip wasn't fixed until cataclysm so if it's a blizzlike bug there's no need to change it (because it involves a custom patch)

**Comment by Stifler82** [comment]
> But how would I know the video would explain it? It is not stated anywhere. Anyhow, I can watch it later :)


Tests Performed:
-Compiled without errors
**-Tested in game and working
https://youtu.be/5X3usUTZiJI (New! Video! English Version) (Update 05/February/2021 )**
-Tested on Windows 10 (64)


How to Test the Changes:
Check the amount of life subtracted from the spell and the fizzle.
Check the amount of mana returned
**https://youtu.be/5X3usUTZiJI My Test and Investigation about the Spell.**

**Comment by Keader** [comment]
I remember to check it in TC long time ago and compare with videos/sniffs/patch notes...
I'm pretty sure that is blizzlike there, maybe you can check scripts in:
https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Spells/spell_warlock.cpp#L888-L920
Btw you are missing penalty too.

**Comment by FrancescoBorzi** [comment]
@azerothcore/testers @pklloveyou can you give another try to this now that it's complete?

**Comment by Stifler82** [comment]
> I remember to check it in TC long time ago and compare with videos/sniffs/patch notes...
> I'm pretty sure that is blizzlike there, maybe you can check scripts in:
> https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Spells/spell_warlock.cpp#L888-L920
> Btw you are missing penalty too.
@Keader 
If you have watched the video of my Fix, you will notice that the spell works and does everything it has to do, it does not miss anything, so I have no idea what penalty you are talking about, can you be clearer please?

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by Keader** [comment]
> > I remember to check it in TC long time ago and compare with videos/sniffs/patch notes...
> > I'm pretty sure that is blizzlike there, maybe you can check scripts in:
> > https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Spells/spell_warlock.cpp#L888-L920
> > Btw you are missing penalty too.
> 
> @Keader
> If you have watched the video of my Fix, you will notice that the spell works and does everything it has to do, it does not miss anything, so I have no idea what penalty you are talking about, can you be clearer please?

If a level 80 use low ranking spells, penalty must be applied, to make player get less mana.

**Comment by Stifler82** [comment]
> > If you have watched the video of my Fix, you will notice that the spell works and does everything it has to do, it does not miss anything, so I have no idea what penalty you are talking about, can you be clearer please?
> 
> If a level 80 use low ranking spells, penalty must be applied, to make player get less mana.

Does every kind of spell have this effect or just this spell (Life Tap)?
@Keader

**Comment by FrancescoBorzi** [comment]
> > > I remember to check it in TC long time ago and compare with videos/sniffs/patch notes...
> > > I'm pretty sure that is blizzlike there, maybe you can check scripts in:
> > > https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Spells/spell_warlock.cpp#L888-L920
> > > Btw you are missing penalty too.
> > 
> > 
> > @Keader
> > If you have watched the video of my Fix, you will notice that the spell works and does everything it has to do, it does not miss anything, so I have no idea what penalty you are talking about, can you be clearer please?
> 
> If a level 80 use low ranking spells, penalty must be applied, to make player get less mana.

Does this PR make it worse in this sense? From what is my understanding, I think we can merge it as it's already an improvement.

**Comment by Keader** [comment]
> > > If you have watched the video of my Fix, you will notice that the spell works and does everything it has to do, it does not miss anything, so I have no idea what penalty you are talking about, can you be clearer please?
> > 
> > 
> > If a level 80 use low ranking spells, penalty must be applied, to make player get less mana.
> 
> Does every kind of spell have this effect or just this spell (Life Tap)?
> @Keader

A lot of spells has this behavior.
Search "penalty" string in core, you find where it's used.



> > > > I remember to check it in TC long time ago and compare with videos/sniffs/patch notes...
> > > > I'm pretty sure that is blizzlike there, maybe you can check scripts in:
> > > > https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/scripts/Spells/spell_warlock.cpp#L888-L920
> > > > Btw you are missing penalty too.
> > > 
> > > 
> > > @Keader
> > > If you have watched the video of my Fix, you will notice that the spell works and does everything it has to do, it does not miss anything, so I have no idea what penalty you are talking about, can you be clearer please?
> > 
> > 
> > If a level 80 use low ranking spells, penalty must be applied, to make player get less mana.
> 
> Does this PR make it worse in this sense? From what is my understanding, I think we can merge it as it's already an improvement.

is a improvement

**Comment by FrancescoBorzi** [comment]
Marking as [To Be Merged] as test evidence has been provided in the videos.

This might not be perfect yet but it's already an improvement of what we currently have.

**Comment by BarbzYHOOL** [comment]
@keader @stifler82 thank you guys

**Pull request merged**